Issue 159 - syncing script updates

Bankso opened the pull request:
Added updates to syncing scripts to align with new CCKP table schemas

Bankso committed:
Update sync_publications.py

Drop unused columns and ignore errors when dropping columns

Bankso committed:
Update sync_datasets.py

Added new columns for CCKP table schema
Added code to drop unused columns
Added logic to parse link and alias for Synapse-related patterns. Script will preferentially copy an existing Synapse link from the link column to the synapseLink column. If an appropriate link doesn't exist, the script will check if alias is a synId. If true, a new Synapse URL will be constructed and listed in synapseLink

Bankso committed:
Update sync_tools.py

Updated columns to match UNION table content and CCKP table schema
Added code to drop unused columns
Added logic to parse multiple tool links for Synapse-related patterns. Script will copy any existing Synapse links from the link columns to the synapseLink column. This can generate a list of links, in format [Link](Synapse URL)

Bankso committed:
Update sync_education.py

Added columns for CCKP table schema
Added code to remove unused columns
Added logic to parse link and alias for Synapse-related patterns. Script will preferentially check if alias is a synId. If true, a new Synapse URL will be constructed and listed in synapseLink. If the alias is not a synId, the script will check if the Resource Link is a Synapse URL and copy it to the synapseLink column, if true. Resource links will have the format [synId](URL) for alias match and [Link](URL) if a URL match

vpchung commented on portal_tables/sync_datasets.py:
```diff
@@ -16,15 +16,35 @@ def add_missing_info(
+                is_in_synapse = True
+        if is_in_synapse:
+            datasets.at[_, "synapseLink"] = datasets.at[_, "link"]
+        else:
```
I think you could simplify this a bit, by using `any()`:

```suggestion
        if any(url_pattern.match(s) for s in row["link"].split("](")):
            datasets.at[_, "synapseLink"] = datasets.at[_, "link"]
        else:
```

Or maybe even simpler with this:

```suggestion
        if "synapse.org" in row["link"]:
            datasets.at[_, "synapseLink"] = datasets.at[_, "link"]
        else:
```

Unless there are downstream implications that require using regex instead?

Bankso replied:
> Unless there are downstream implications that require using regex instead?

This is a good question. I wanted to make sure we had flexibility to identify any URL pattern of interest, in case we want to do a similar extraction for other databases, but I think that can be achieved using your suggestion.

vpchung replied:
Oh, that makes sense! I like it.

I was actually thinking that since these same set of lines are used across multiple scripts, that we could extract them out into the `utils` module as a function.  We could probably set the pattern (and maybe new colname) as the parameters, thus giving you that flexibility in URL pattern.

Bankso replied:
That sounds great, I've thought about that several times (and I know you've done that for some functions previously) so I'm glad you brought it up!

vpchung commented on portal_tables/sync_datasets.py:
```diff
@@ -16,15 +16,35 @@ def add_missing_info(
+                if re.match(alias_pattern, a):
+                    syn_link = "".join(["https://www.synapse.org/Synapse:", a])
+                    formatted_syn_link = "".join(["[", a, "](", syn_link, ")"])
+                    syn_link_list.append(formatted_syn_link)
```
A list comprehension and f-string could be helpful here!

```suggestion
            syn_link_list = [
                f"[{a}](https://www.synapse.org/Synapse:{a})"
                for a in alias_list
                if re.match(alias_pattern, a)
            ]
```

vpchung replied:
Also, not sure if you already knew this, but since you're using `re.match()` there's no need to include the `^` anchor in `alias_pattern`!  Since `re.match` looks for a pattern match at the start of the string already.  If you were using `re.search`, then the `^` would be necessary.

Bankso replied:
Love the list comprehension, thanks for pointing that out!

> there's no need to include the ^ anchor in `alias_pattern`! Since `re.match` looks for a pattern match at the start of the string already. If you were using `re.search`, then the ^ would be necessary.

I think I started with `re.search` and didn't change the pattern, but that makes sense!

vpchung replied:
It's always good to be explicit anyway, so I'd just keep the regex pattern as-is :-)

Just wanted to share as a general FYI!

vpchung commented on portal_tables/sync_datasets.py:
```diff
@@ -48,12 +68,17 @@ def add_missing_info(
     """Clean up the table one final time."""
 
+    df["DatasetGrantNumber"] = df["GrantViewKey"]
+    df["DatasetPubmedId"] = df["PublicationViewKey"]
```
nit: instead of creating duplicate columns, you can rename the colnames instead:

```suggestion
    df = df.rename(columns={
        "DatasetGrantNumber": "GrantViewKey",
        "DatasetPubmedId": "PublicationViewKey"
    }]
```

Bankso replied:
Good point, I like that better!

vpchung commented on portal_tables/sync_datasets.py:
```diff
@@ -48,12 +68,17 @@ def add_missing_info(
 
+    df["DatasetGrantNumber"] = df["GrantViewKey"]
+    df["DatasetPubmedId"] = df["PublicationViewKey"]
+    df = df.drop(["GrantViewKey", "PublicationViewKey", "StudyKey"], errors="ignore")
```
nit: since the function is returning `df` with a select list of columns (and in the order we want), no need to drop any columns here :-)

Bankso replied:
> since the function is returning df with a select list of columns (and in the order we want), no need to drop any columns here :-)

I actually noticed that when I was testing and forgot to adjust the code. My brain wasn't interpreting `return df[col_order]` as a subsetting action for some reason, so thank you for pointing this out!

vpchung commented on portal_tables/sync_education.py:
```diff
@@ -5,12 +5,48 @@
+    df["ResourcePubmedId"] = df["PublicationViewKey"]
+    df["ResourceDatasetAlias"] = df["DatasetViewKey"]
+    df["ResourceToolLink"] = df["ToolViewKey"]
+    df = df.drop(["GrantViewKey", "PublicationViewKey", "DatasetViewKey", "ToolViewKey", "StudyKey"], errors="ignore")
```
Same comment as above about renaming and dropping columns.

vpchung commented on portal_tables/sync_publications.py:
```diff
@@ -49,6 +49,9 @@ def add_missing_info(
     """Clean up the table one final time."""
 
+    df["Publication Grant Number"] = df["GrantView Key"]
+    df = df.drop(["PublicationView_id", "GrantView Key", "Study Key"], errors="ignore")
```
Same comment as above about renaming and dropping columns.

vpchung commented on portal_tables/sync_tools.py:
```diff
@@ -5,32 +5,49 @@
+    df["ToolGrantNumber"] = df["GrantViewKey"]
+    df["ToolPubmedId"] = df["PublicationViewKey"]
+    df["ToolDatasets"] = df["DatasetViewKey"]
+    df = df.drop(["ToolView_id", "GrantViewKey", "PublicationViewKey", "DatasetViewKey", "StudyKey"], errors="ignore")
```
Same comment as above about renaming and dropping columns.

vpchung commented on portal_tables/sync_education.py:
```diff
@@ -5,12 +5,48 @@
+                    s_match = re.match(url_pattern, s)
+                    is_in_synapse = True if s_match else None
+                    if is_in_synapse:
+                        education.at[_, "synapseLink"] = "".join(["[Link](", education.at[_, "ResourceLink"], ")"])
```
Looks like `synapseLink` could potentially be assigned a new value every time the loop iterates through `alias_list`. Maybe move the assignment outside of this loop to ensure it only gets populated once?, e.g.

```python
        for a in alias_list:
             ...
             ...
        education.at[_, "synapseLink"] = ...
```

Bankso replied:
I see what you're saying! I was setting this up to preferentially convert a `synId` into a `synapseLink` and, since their should only ever be one alias in `Resource Alias`, I didn't address the implication of using a `list` in this context. Reading it now, I realize I was just trying very hard to reuse the link extraction/assignment structure in the other scripts. Lots of room to simplify!

vpchung commented on portal_tables/sync_education.py:
```diff
@@ -5,12 +5,48 @@
+                    s_match = re.match(url_pattern, s)
+                    is_in_synapse = True if s_match else None
+                    if is_in_synapse:
+                        education.at[_, "synapseLink"] = "".join(["[Link](", education.at[_, "ResourceLink"], ")"])
```
At a quick glance, there is a lot going on here 😅  I think using list comprehensions and f-strings could help improve the readability.

Bankso replied:
Sounds good, I will adjust it!

vpchung commented on portal_tables/sync_tools.py:
```diff
@@ -5,32 +5,49 @@
+            if s_match:
+                print(s)
+                synapse_links.append("".join(["[Link](", s , ")"]))
+        tools.at[_, "synapseLink"] = ", ".join(list(set(synapse_links)))
```
No need to type cast back to a list, since sets are also iterable :-)

```suggestion
        tools.at[_, "synapseLink"] = ", ".join(set(synapse_links))
```

Bankso replied:
Makes sense!

vpchung reviewed (changes requested):
Nice!  I made some suggestions that I think could help improve the readability.

Bankso committed:
Update portal_tables/sync_datasets.py

Co-authored-by: Verena Chung <9377970+vpchung@users.noreply.github.com>

Bankso committed:
Update portal_tables/sync_datasets.py

Co-authored-by: Verena Chung <9377970+vpchung@users.noreply.github.com>

Bankso committed:
Update portal_tables/sync_datasets.py

Co-authored-by: Verena Chung <9377970+vpchung@users.noreply.github.com>

Bankso committed:
Update portal_tables/sync_tools.py

Co-authored-by: Verena Chung <9377970+vpchung@users.noreply.github.com>

Bankso committed:
Update sync_datasets.py

rename columns instead of adding and dropping

Bankso committed:
Update sync_education.py

rename columns instead of adding and dropping

Bankso committed:
Update sync_publications.py

rename columns instead of adding and dropping

Bankso committed:
Update sync_tools.py

rename columns instead of adding and dropping

Bankso committed:
Update sync_education.py

Condense alias and url extraction

Bankso committed:
Update sync_education.py

adjust order to prefer building url from alias and reformat list comprehension for url capture

Bankso committed:
Update sync_education.py

Cast entries to pandas series to enable conversion to lists. addresses error `AttributeError: 'DataFrame' object has no attribute 'str'.` when trying to run function utils.convert_to_stringlist(). Functions expects series but other changes were resulting in a dataframe object being passed instead

Bankso committed:
Update sync_publications.py

adjust df.rename input

Bankso committed:
Update sync_tools.py

adjust df.rename input
